source/opt/instruction_list.cpp

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "instruction_list.h"
+
+namespace spvtools {
+namespace ir {
+
+InstructionList::~InstructionList() {
+  while (!empty()) {
+    Instruction* inst = &front();
+    inst->RemoveFromList();
+    delete inst;
+  }
+}
+
+InstructionList::iterator InstructionList::iterator::InsertBefore(
+    std::vector<std::unique_ptr<Instruction>>* list) {
```

Comment:
Would it make sense to have an ::InsertBefore() that accepted an InstructionList?

Reply:
There is one that is part of the base class.  I wrote this one because it existed for the old iterator type.  As we change the code that uses it we can phase it out.  I will add a comment to say it is deprecated, and new code should not use it.